activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -516,6 +519,11 @@ def find_by_scan(*args)
             load_target.select { |r| ids.include?(r.id.to_s) }
           end
         end
+
+        def index_in_target(record)
+          @target_index_map ||= @target.each_with_object({}).with_index { |(r, memo), i| memo[r] ||= i }
```

Comment:
The code this was replacing was `@target.index(record)` which would return the first index for duplicate record so this implementation keeps that behavior.